File: katas/4 kyu/permutations/solution.py
# permutations
import codewars_test as test
# -------------------------------- MY SOLUTION ------------------------------- #
import itertools as iter
def permutations(s):
  return list(set(["".join(x) for x in iter.permutations(s)]))
# ------------------------------ CLEVER SOLUTION ----------------------------- #
import itertools
def permutations(string):
  return list("".join(p) for p in set(itertools.permutations(string)))
# ----------------------------------- TEST ----------------------------------- #
@test.describe("Basic Tests")
def basic_tests():
    
  @test.it("Basic Tests")
  def basic_tests():
    test.assert_equals(sorted(permutations('a')), ['a']);
    test.assert_equals(sorted(permutations('ab')), ['ab', 'ba'])
    test.assert_equals(sorted(permutations('aabb')), ['aabb', 'abab', 'abba', 'baab', 'baba', 'bbaa'])

File: katas/4 kyu/permutations/solution.php
<!-- permutations -->
<?php
use PHPUnit\Framework\TestCase;
/* ------------------------------- MY SOLUTION ------------------------------ */
function permutations(string $str): array
{
  if (strlen($str) < 2) {
    return [$str];
  }

  $result = [];
  $stop = strlen($str) - 1;
  for ($i = 0; $i <= $stop; $i++) {
    $substr = substr($str, 0, $i) . substr($str, $i + 1);
    foreach (permutations($substr) as $per) {
      $result[] = $str[$i] . $per;
    }
  }
  return array_unique($result);
}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
/* ---------------------------------- TEST ---------------------------------- */
final class PermutationsTest extends TestCase {
  private function assertEquivalent(array $expected, array $actual, string $msg = '') {
    sort($expected);
    sort($actual);
    $this->assertSame($expected, $actual, $msg);
  }
  public function testDescriptionExamples() {
    $this->assertEquivalent(['a'], permutations('a'));
    $this->assertEquivalent(['ab', 'ba'], permutations('ab'));
    $this->assertEquivalent(['aabb', 'abab', 'abba', 'baab', 'baba', 'bbaa'], permutations('aabb'));
  }
}